ui/components/app/signature-request/signature-request.component.js

```diff
@@ -45,9 +45,72 @@ export default class SignatureRequest extends PureComponent {
       hardwareWalletRequiresConnection,
     } = this.props;
     const { address: fromAddress } = fromAccount;
-    const { message, domain = {} } = JSON.parse(data);
+    const { message, domain = {}, types, primaryType } = JSON.parse(data);
     const { metricsEvent } = this.context;
 
+    const solidityTypes = [
+      'bool',
+      'int',
+      'fixed',
+      'address',
+      'bytes',
+      'string',
+    ];
+    const sanitizeMessage = (msg, baseType) => {
+      let baseTypeType;
+      if (types) {
+        baseTypeType = types[baseType];
+        if (!baseTypeType) {
+          return msg;
+        }
+      } else {
+        return msg;
+      }
+
+      const sanitizedMessage = {};
+      const msgKeys = Object.keys(msg);
+      msgKeys.forEach((msgKey) => {
+        const definedType = Object.values(baseTypeType).find(
+          (ptt) => ptt.name === msgKey,
+        );
+
+        if (definedType) {
+          // key has a type. check if the definedType is also a type
+          const nestedTypeType = types[definedType.type.replace('[]', '')];
+          if (nestedTypeType) {
+            if (definedType.type.indexOf('[]') > 0) {
+              // nested array
+              const sanitizedArrayMessage = {};
+              const definedArrayType = definedType.type.replace('[]', '');
+              const arrayMsg = msg[msgKey];
+              arrayMsg.forEach((msgArray, arrIndex) => {
+                sanitizedArrayMessage[arrIndex] = sanitizeMessage(
+                  msgArray,
+                  definedArrayType,
+                );
+              });
+              sanitizedMessage[msgKey] = sanitizedArrayMessage;
```

Comment:
This seems overly complex. This can be done more simply with `.map`.

e.g.

```suggestion
              sanitizedMessage[msgKey] = msg[msgKey].map((value) => sanitizeMessage(value, definedArrayType));
```